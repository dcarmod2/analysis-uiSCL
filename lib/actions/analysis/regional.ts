/** Actions for regional analysis */
import get from 'lodash/get'
import sortBy from 'lodash/sortBy'
import {createAction} from 'redux-actions'

import {API} from 'lib/constants'
import R5Version from 'lib/modules/r5-version'
import createGrid from 'lib/utils/create-grid'
import predictJobTimeRemaining from 'lib/utils/predict-job-time-remaining'

import fetch from '../fetch'
import fetchSignedS3Url from '../fetch-signed-s3-url'

const REGION_URL = API.Region
const REGIONAL_URL = API.Regional

export const setRegionalAnalysis = createAction('set regional analysis')
export const setRegionalAnalyses = createAction('set regional analyses')
export const setActiveRegionalJobs = createAction('set active regional jobs')

export const load = (regionId) =>
  fetch({
    url: `${REGION_URL}/${regionId}/regional`,
    next(response) {
      const analyses = sortBy(response.value, (a) => -a.createdAt) // newest at the top
      return [
        setRegionalAnalyses(analyses),
        R5Version.actions.setUsedVersions(
          analyses.map((a) => ({
            name: a.name,
            version: a.workerVersion
          }))
        )
      ]
    }
  })

// Show time if...at least one task is complete AND either the regional analysis
// was created more than five minutes ago or there is more than one acitve worker.
function getJobStatus(job) {
  if (job.complete === job.total) return 'assembling results...'
  if (job.complete === 0) return 'starting cluster...'
  return predictJobTimeRemaining(job)
}

export const loadActiveRegionalJobs = (regionId) =>
  fetch({
    url: `${REGION_URL}/${regionId}/regional/running`,
    next: (res) =>
      setActiveRegionalJobs(
        (res.value || []).map((j) => ({
          ...j,
          statusText: getJobStatus(j)
        }))
      )
  })

export const setActiveRegionalAnalyses = createAction(
  'set active regional analyses'
)
export const setRegionalAnalysisGrid = createAction(
  'set regional analysis grid'
)
const setRegionalAnalysisOriginLocally = createAction(
  'set regional analysis origin'
)

export const loadRegionalAnalysisGrid = (
  analysis,
  cutoff,
  percentile,
  pointSetId
) => async (dispatch, getState) => {
  const {grids} = getState().regionalAnalyses

  // Check if the analysis needs a cutoff and percentile
  const hasMultiCutoff =
    Array.isArray(analysis.cutoffsMinutes) &&
    Array.isArray(analysis.travelTimePercentiles)
  if (
    hasMultiCutoff &&
    (!Number.isInteger(cutoff) || !Number.isInteger(percentile))
  ) {
    console.error(
      'Multi-dimensional analysis needs cutoff and percentile',
      cutoff,
      percentile
    )
    return
  }

  if (
    !grids.find(
      (g) =>
        g.analysisId === analysis._id &&
        g.cutoff === cutoff &&
        g.percentile === percentile &&
        g.pointSetId === pointSetId
    )
  ) {
    const rawGrid = await dispatch(
      fetchSignedS3Url(
        `${REGIONAL_URL}/${analysis._id}/grid/grid?cutoff=${cutoff}&percentile=${percentile}&destinationPointSetId=${pointSetId}`
      )
    )

    // Pass into the store
    dispatch(
      setRegionalAnalysisGrid({
        // Parse the headers and create a usable grid.
        ...createGrid(rawGrid),
        // Assign the regional analysis _id and display parameters
        analysisId: analysis._id,
        cutoff: cutoff || get(analysis, 'cutoffMinutes'),
        percentile: percentile || get(analysis, 'travelTimePercentile'),
        pointSetId
      })
    )
  }
}

export const createRegionalAnalysis = (settings) =>
  fetch({
    options: {method: 'POST', body: settings},
    url: REGIONAL_URL
  })

const deleteRegionalAnalysisLocally = createAction('delete regional analysis')
export const deleteRegionalAnalysis = (analysisId) => (dispatch) => {
  // Run local delete first so it seems snappier. The worst that will happen is
  // that a regional analysis will pop back up in a few seconds when we refresh
  // regional analyses.
  dispatch(deleteRegionalAnalysisLocally(analysisId))

  return dispatch(
    fetch({
      url: `${REGIONAL_URL}/${analysisId}`,
      options: {
        method: 'DELETE'
      },
      next: () => deleteRegionalAnalysisLocally(analysisId)
    })
  )
}

export function downloadGridFromS3(url) {
  return fetch({
    url,
    next: (err, response) => {
      if (err) window.alert(err)
      else window.open(response.url)
    }
  })
}

/**
 * NB: THIS FUNCTIONALITY IS CURRENTLY DISABLED.
 *
 * Set the origin for showing a bootstrapped sampling distribution of travel
 * time.
 */
export const setRegionalAnalysisOrigin = (opts) => {
  if (!opts) return setRegionalAnalysisOriginLocally(null)
  else return setRegionalAnalysisOriginLocally(opts.lonlat)
}

/**
 * Update a regional analysis
 */
export const updateRegionalAnalysis = (regionalAnalysis) =>
  fetch({
    options: {
      method: 'put',
      body: regionalAnalysis
    },
    url: `${REGIONAL_URL}/${regionalAnalysis._id}`,
    next: (response) => setRegionalAnalysis(response.value)
  })
