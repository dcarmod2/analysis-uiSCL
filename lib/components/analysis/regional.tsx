import {
  Box,
  Select,
  Stack,
  Flex,
  useDisclosure,
  Menu,
  MenuButton,
  Button,
  MenuList,
  MenuItem
} from '@chakra-ui/core'
import {
  faChevronLeft,
  faDownload,
  faTrash
} from '@fortawesome/free-solid-svg-icons'
import find from 'lodash/find'
import get from 'lodash/get'
import omit from 'lodash/omit'
import dynamic from 'next/dynamic'
import React, {useCallback} from 'react'
import {useDispatch, useSelector} from 'react-redux'

import {setSearchParameter} from 'lib/actions'
import {
  deleteRegionalAnalysis,
  updateRegionalAnalysis
} from 'lib/actions/analysis/regional'
import fetchAction from 'lib/actions/fetch'
import {API} from 'lib/constants'
import useControlledInput from 'lib/hooks/use-controlled-input'
import message from 'lib/message'
import selectActiveAnalysis from 'lib/selectors/active-regional-analysis'
import selectCutoff from 'lib/selectors/regional-display-cutoff'
import selectPercentile from 'lib/selectors/regional-display-percentile'
import selectPointset from 'lib/selectors/regional-display-destination-pointset'
import downloadObjectAsJson from 'lib/utils/download-json'

import {ConfirmDialog} from '../confirm-button'
import Editable from '../editable'
import IconButton from '../icon-button'
import RunningAnalysis from '../running-analysis'

import ProfileRequestDisplay from './profile-request-display'
import Icon from '../icon'

const AggregationArea = dynamic(() => import('../map/aggregation-area'), {
  ssr: false
})
const AnalysisBounds = dynamic(() => import('../map/analysis-bounds'), {
  ssr: false
})
const DotMap = dynamic(
  () => import('lib/modules/opportunity-datasets/components/dotmap'),
  {ssr: false}
)
const RegionalLayer = dynamic(() => import('../map/regional'), {
  ssr: false
})
const RegionalResults = dynamic(() => import('./regional-results'), {
  ssr: false
})

// Ensure valid analysis name
const nameIsValid = (s) => s && s.length > 0

// Get full qualifier for opportunity datasets
const getFullODName = (od) => `${od?.sourceName}: ${od?.name}`

// Type to title
const csvResultsTypeToTitle = {
  ACCESS: 'Access CSV',
  PATHS: 'Paths CSV',
  TIMES: 'Times CSV'
}

export default function Regional(p) {
  const deleteDisclosure = useDisclosure()
  const isComplete = !p.job
  const analysis = useSelector(selectActiveAnalysis)
  const dispatch = useDispatch<any>()
  const cutoffsMinutes = analysis.cutoffsMinutes
  const percentiles = analysis.travelTimePercentiles

  const onChangeCutoff = useCallback(
    (v) => dispatch(setSearchParameter('cutoff', v)),
    [dispatch]
  )
  const cutoffInput = useControlledInput({
    onChange: onChangeCutoff,
    parse: parseInt,
    value: useSelector(selectCutoff)
  })

  const onChangePercentile = useCallback(
    (v) => dispatch(setSearchParameter('percentile', v)),
    [dispatch]
  )
  const percentileInput = useControlledInput({
    onChange: onChangePercentile,
    parse: parseInt,
    value: useSelector(selectPercentile)
  })

  const activePointSet = useSelector(selectPointset)
  const activePointSetId = get(activePointSet, '_id')
  const onChangeDestinationPointSet = useCallback(
    (v) => dispatch(setSearchParameter('destinationPointSetId', v)),
    [dispatch]
  )
  const destinationPointSetInput = useControlledInput({
    onChange: onChangeDestinationPointSet,
    value: activePointSetId
  })

  async function _remove() {
    await dispatch(deleteRegionalAnalysis(analysis._id))
    return dispatch(setSearchParameter('analysisId'))
  }

  async function _downloadCSVResults(type) {
    const url = await dispatch(
      fetchAction({url: `${API.Regional}/${analysis._id}/csv/${type}`})
    )
    window.open(url)
  }

  async function _downloadRequestJSON() {
    const data = await dispatch(
      fetchAction({url: `${API.Regional}/${analysis._id}`})
    )
    downloadObjectAsJson({data, filename: analysis.name + '.json'})
  }

  /**
   * Perform an authenticated fetch to get a presigned URL to download a grid
   * from S3, then download it. Pass in the path to fetch.
   */
  async function _downloadProjectGIS(e) {
    e.preventDefault()
    const value = await dispatch(
      fetchAction({
        url: `${API.Regional}/${analysis._id}/grid/tiff?cutoff=${cutoffInput.value}&percentile=${percentileInput.value}&destinationPointSetId=${activePointSetId}`
      })
    )
    window.open(get(value, 'url'))
  }

  return (
    <>
      {!isComplete ? (
        <AnalysisBounds analysis={analysis} />
      ) : (
        <>
          <DotMap />
          <RegionalLayer />
          <AggregationArea />
        </>
      )}

      {deleteDisclosure.isOpen && (
        <ConfirmDialog
          action='Delete regional analysis'
          description='Are you sure you would like to delete this analysis?'
          onClose={deleteDisclosure.onClose}
          onConfirm={_remove}
        />
      )}

      <Flex
        align='center'
        borderBottom='1px solid #E2E8F0'
        className={p.saveInProgress ? 'disableAndDim' : ''}
        p={2}
        width='320px'
      >
        <IconButton
          icon={faChevronLeft}
          label='All regional analyses'
          onClick={() => dispatch(setSearchParameter('analysisId'))}
        />

        <Box flex='1' fontSize='xl' fontWeight='bold' ml={2} overflow='hidden'>
          <Editable
            isValid={nameIsValid}
            onChange={(name) =>
              dispatch(updateRegionalAnalysis({...analysis, name}))
            }
            value={analysis.name}
          />
        </Box>

        <Flex>
          <IconButton
            icon={faTrash}
            label={message('analysis.deleteRegionalAnalysis')}
            onClick={deleteDisclosure.onOpen}
            variantColor='red'
          />
        </Flex>
      </Flex>

      {p.job && <RunningAnalysis job={p.job} />}

      <Stack spacing={4} px={4} py={4}>
        {Array.isArray(analysis.destinationPointSetIds) && (
          <Select {...destinationPointSetInput}>
            {analysis.destinationPointSetIds.map((id) => (
              <option key={id} value={id}>
                {getFullODName(find(p.opportunityDatasets, ['_id', id]))}
              </option>
            ))}
          </Select>
        )}

        <Stack isInline>
          {Array.isArray(cutoffsMinutes) && (
            <Select {...cutoffInput}>
              {cutoffsMinutes.map((m) => (
                <option key={m} value={m}>
                  {m} minutes
                </option>
              ))}
            </Select>
          )}

          {Array.isArray(percentiles) && (
            <Select {...percentileInput}>
              {percentiles.map((p) => (
                <option key={p} value={p}>
                  {p}th percentile
                </option>
              ))}
            </Select>
          )}
        </Stack>

        {isComplete && (
          <Menu>
            <MenuButton as={Button} {...{variant: 'outline'}}>
              <Icon icon={faDownload} />
              &nbsp;&nbsp;Export data
            </MenuButton>
            <MenuList>
              <MenuItem onClick={_downloadProjectGIS}>GeoTIFF</MenuItem>
              <MenuItem onClick={_downloadRequestJSON}>
                Scenario and modification JSON
              </MenuItem>
              {Object.keys(analysis.resultStorage || {}).map((type) => (
                <MenuItem key={type} onClick={() => _downloadCSVResults(type)}>
                  {csvResultsTypeToTitle[type]}
                </MenuItem>
              ))}
            </MenuList>
          </Menu>
        )}
      </Stack>

      <ProfileRequestDisplay
        bundleId={analysis.bundleId}
        profileRequest={{
          ...omit(analysis, 'request'),
          ...analysis.request
        }}
        projectId={analysis.projectId}
      />

      {isComplete && (
        <RegionalResults
          analysis={analysis}
          analysisId={analysis._id}
          regionId={analysis.regionId}
          {...p}
        />
      )}
    </>
  )
}
