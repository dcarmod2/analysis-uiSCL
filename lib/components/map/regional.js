import React from 'react'
import {useSelector} from 'react-redux'

import selectDisplayGrid from 'lib/selectors/regional-display-grid'
import selectDisplayScale from 'lib/selectors/regional-display-scale'
import createDrawTile from 'lib/utils/create-draw-tile'

import Gridualizer from './gridualizer'

/**
 * A map layer showing a Regional comparison
 */
let key = 0
export default React.memo(function Regional() {
  const grid = useSelector(selectDisplayGrid)
  const {error, colorizer} = useSelector(selectDisplayScale) || {}

  if (!grid || !colorizer || error) return null

  // Function is cheap. It just wraps draw tile with the necessary parts
  const drawTile = createDrawTile({colorizer, grid})

  // Remount every time there is a change to drawTile
  return <Gridualizer key={key++} drawTile={drawTile} zIndex={300} />
})
