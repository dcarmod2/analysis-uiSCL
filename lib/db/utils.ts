import mapValues from 'lodash/mapValues'
import {ObjectID} from 'mongodb'

export function serializeCollection(obj) {
  if (obj instanceof Date) return obj.toISOString()
  if (obj instanceof ObjectID) return obj.toString()
  if (Array.isArray(obj)) return obj.map(serializeCollection)
  if (typeof obj === 'object') return mapValues(obj, serializeCollection)
  return obj
}
