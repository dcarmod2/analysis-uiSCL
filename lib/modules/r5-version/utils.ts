import {MINIMUM_R5_VERSION, VERSION_PARSE_REGEX} from './constants'

/**
 * Convert an r5 version to a number for comparing
 */
export function versionToNumber(version: string): number {
  const versionMatch = (version || '').match(VERSION_PARSE_REGEX)
  if (versionMatch) {
    const [, major, minor, patch, commit] = versionMatch

    // NB Assumes no major, minor, patch, or commit version exceeds 100
    return (
      parseInt(major) * 10000 +
      parseInt(minor) * 100 +
      parseInt(patch) +
      parseInt(commit || '0') / 100
    )
  }

  return 0
}

/**
 * Check if the r5 version is older than a given minimum.
 */
export function workerVersionInRange(
  version: string,
  min: string = MINIMUM_R5_VERSION,
  max = 'v1000.0.0'
): boolean {
  const versionNumber = versionToNumber(version)
  return (
    versionNumber >= versionToNumber(min) &&
    versionNumber <= versionToNumber(max)
  )
}

/**
 * Check if the r5 version is a patch version or in a given range
 */
export function workerVersionTestOrInRange(
  version: string,
  min: string,
  max?: string
): boolean {
  const isTestVersion = versionToNumber(version) % 1 !== 0
  return isTestVersion || workerVersionInRange(version, min, max)
}
