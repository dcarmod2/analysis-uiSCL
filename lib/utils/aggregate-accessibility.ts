import get from 'lodash/get'
import range from 'lodash/range'

const N_HISTOGRAM_BINS = 15

/**
 * Compute the accessibility, exported for use in comparison weighted average
 * accessibility.
 * @param weights a grid containing the weights (e.g. population)
 * @param aggregationArea a grid containing the aggregation area, with the
 *        percentage the aggregation area overlaps each cell, scaled between 0
 *        (no overlap) and 100,000 (full coverage)
 * @param accessibility a grid containing the accessibility (i.e. the regional
 *        analysis results)
 */
export default function getAggregateAccessibility(
  accessibility,
  aggregationArea,
  weights
) {
  const aa = get(aggregationArea, 'grid')
  if (weights == null || aa == null || accessibility == null) {
    return
  }

  if (accessibility.zoom !== weights.zoom || accessibility.zoom !== aa.zoom) {
    console.error('Zoom levels of all grids in weighted average must match!')
  }

  // find max and min value for histogram
  let maxAccessibility = 0
  let minAccessibility = Infinity

  // Pass one: find min and max from aggregationArea
  // aggregationArea is probably the smallest grid and is guaranteed to contain
  // all pixels
  for (let y = 0, pixel = 0; y < aa.height; y++) {
    // Transform to coordinates in other grid; all grids have same zoom, so
    // translate based on the difference in where the edges of the grid are
    const accessY = y + aa.north - accessibility.north
    for (let x = 0; x < aa.width; x++, pixel++) {
      if (aa.data[pixel] > 0) {
        const accessX = x + aa.west - accessibility.west
        const accessibilityThisCell = accessibility.contains(accessX, accessY)
          ? accessibility.data[accessY * accessibility.width + accessX]
          : 0

        minAccessibility = Math.min(accessibilityThisCell, minAccessibility)
        maxAccessibility = Math.max(accessibilityThisCell, maxAccessibility)
      }
    }
  }

  const spread = maxAccessibility - minAccessibility
  const bins = range(N_HISTOGRAM_BINS).map(() => 0)

  let sumOfWeights = 0
  // weight and accessibility value for each cell, used to compute percentiles
  const weightedAccessibility = []

  // pass two: accumulate into bins
  for (let y = 0, pixel = 0; y < aa.height; y++) {
    const accessY = y + aa.north - accessibility.north
    const weightY = y + aa.north - weights.north
    for (let x = 0; x < aa.width; x++, pixel++) {
      if (aa.data[pixel] > 0) {
        const accessX = x + aa.west - accessibility.west
        const accessibilityThisCell = accessibility.contains(accessX, accessY)
          ? accessibility.data[accessY * accessibility.width + accessX]
          : 0

        let binIndex = Math.floor(
          ((accessibilityThisCell - minAccessibility) / spread) *
            N_HISTOGRAM_BINS
        )

        if (binIndex > N_HISTOGRAM_BINS) console.error('invalid bin!')

        // avoid putting the highest accessibility cell beyond the histogram
        if (binIndex === N_HISTOGRAM_BINS) binIndex--

        // how much of this cell is within the aggregationArea
        const aggregationAreaWeightThisCell = aa.data[pixel] / 100000

        const weightX = x + aa.west - weights.west
        const weightThisCell = weights.contains(weightX, weightY)
          ? weights.data[weightY * weights.width + weightX]
          : 0

        const weight = aggregationAreaWeightThisCell * weightThisCell
        bins[binIndex] += aggregationAreaWeightThisCell * weightThisCell

        sumOfWeights += weight
        weightedAccessibility.push({
          weight,
          accessibility: accessibilityThisCell
        })
      }
    }
  }

  // Figure out percentiles
  let cumulativeWeight = 0
  // Preinitialize with null so 50th percentile is percentiles[50].
  // There is no zeroth percentile
  const percentiles = [0]
  const weightPerPercentile = sumOfWeights / 100
  let expectedWeight = weightPerPercentile
  weightedAccessibility.sort((a, b) => a.accessibility - b.accessibility)

  // we calculate and return an old-style weighted average for comparison purposes.
  let sumOfWeightedAccessibility = 0

  // We loop over the sorted weighted accessibility values. When we cross a
  // percentile boundary, we push that percentile onto the list
  weightedAccessibility.forEach(({accessibility, weight}) => {
    sumOfWeightedAccessibility += accessibility * weight

    cumulativeWeight += weight

    // while not if, possible to cross many percentiles with one high-density cell
    while (cumulativeWeight > expectedWeight) {
      // There is a percentile value somewhere within the values at this sample
      // We use > above, not >=, so that a percentile value that falls exactly
      // between two accessibility values will get the higher value, e.g. if 10
      // people have accessibility 50 and ten people have accessibility 60, we
      // want 60. This is because in the UI we say "x% of <people> have
      // accessibility y or higher," so we want y to be as high as possible
      // while still being a true statement.
      percentiles.push(accessibility)
      expectedWeight += weightPerPercentile
    }
  })

  // includes values 1 - 99
  // When the resolution of the data is low the last item may only define (say)
  // the 90th percentile, all remaining percentiles are just the max. Also
  // handles the case when there is no accessibility within the aggregation
  // area by returning all zeros.
  while (percentiles.length < 99) {
    percentiles.push(
      weightedAccessibility.length > 0
        ? weightedAccessibility[weightedAccessibility.length - 1].accessibility
        : 0
    )
  }

  const binWidth = spread / N_HISTOGRAM_BINS

  return {
    minAccessibility,
    maxAccessibility,
    percentiles,
    weightedAverage: sumOfWeightedAccessibility / sumOfWeights,
    bins: bins.map((value, idx) => ({
      value,
      min: minAccessibility + idx * binWidth,
      max: minAccessibility + (idx + 1) * binWidth
    }))
  }
}
