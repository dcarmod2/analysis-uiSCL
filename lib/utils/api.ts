// Query values are typed as string OR string[]
export function getQueryAsString(p: string | string[]): string {
  return Array.isArray(p) ? p[0] : p
}

// Get a query as an object
export function getQueryAsObject(
  p: string | string[]
): Record<string, unknown> {
  try {
    return JSON.parse(getQueryAsString(p))
  } catch (e) {
    return {}
  }
}

// Convert an error to POJO for sending over the wire
export function errorToPOJO(e: Error): Record<string, unknown> {
  const obj = {}
  Object.getOwnPropertyNames(e).forEach((k) => {
    obj[k] = e[k]
  })
  return obj
}
