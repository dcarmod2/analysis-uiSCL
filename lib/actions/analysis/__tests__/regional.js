import nock from 'nock'

import {makeMockStore, mockRegionalAnalyses} from 'lib/utils/mock-data'

import * as regional from '../regional'

describe('Actions > Analysis > Regional', () => {
  it('setRegionalAnalyses should work', () => {
    expect(regional.setRegionalAnalyses(mockRegionalAnalyses)).toMatchSnapshot()
  })

  it('setActiveRegionalAnalyses should work', () => {
    expect(regional.setActiveRegionalAnalyses()).toMatchSnapshot()
  })

  it('deleteRegionalAnalysis should work', () => {
    const store = makeMockStore()
    const nockHost = nock('http://localhost')
      .delete('/api/regional/12321a')
      .reply(200, 'deleted')

    return store
      .dispatch(regional.deleteRegionalAnalysis('12321a'))
      .then(() => {
        expect(nockHost.isDone()).toBeTruthy()
      })
  })

  // TODO loadRegionalAnalysisGrids, runAnalysis, deleteRegionalAnalysis
})
