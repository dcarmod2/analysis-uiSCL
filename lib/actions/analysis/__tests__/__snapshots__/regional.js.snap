// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions > Analysis > Regional setActiveRegionalAnalyses should work 1`] = `
Object {
  "type": "set active regional analyses",
}
`;

exports[`Actions > Analysis > Regional setRegionalAnalyses should work 1`] = `
Object {
  "payload": Array [
    Object {
      "_id": "abcd",
      "accessGroup": "accessGroupName",
      "bounds": Object {},
      "bundleId": "1",
      "complete": true,
      "createdAt": "2017-11-08T12:30:50.133Z",
      "createdBy": "person@gmail.com",
      "cutoffMinutes": 60,
      "cutoffsMinutes": Array [
        60,
      ],
      "deleted": false,
      "destinationPointSetIds": Array [
        "Jobs_total",
      ],
      "grid": "Jobs_total",
      "height": 300,
      "name": "ANALYSIS 1",
      "nonce": "12345",
      "north": 20024,
      "projectId": "1",
      "request": Object {
        "accessModes": "WALK",
        "bikeSafe": 1,
        "bikeSlope": 1,
        "bikeSpeed": 4.166666666666667,
        "bikeTime": 1,
        "bounds": Object {
          "east": -76,
          "north": 39,
          "south": 38,
          "west": -77,
        },
        "bundleId": "1",
        "date": "2016-01-16",
        "defaultExpanded": true,
        "directModes": "WALK",
        "egressModes": "WALK",
        "fromLat": 12,
        "fromLon": 34,
        "fromTime": 25200,
        "maxBikeTime": 20,
        "maxRides": 4,
        "maxTripDurationMinutes": 60,
        "maxWalkTime": 20,
        "monteCarloDraws": 200,
        "name": "Profile Request Test",
        "projectId": "1",
        "reachabilityThreshold": 0,
        "regionId": "1",
        "toTime": 32400,
        "transitModes": "BUS,TRAM,RAIL,SUBWAY",
        "travelTimePercentile": 50,
        "variantIndex": 0,
        "walkSpeed": 1.3888888888888888,
        "workerVersion": "v4.0.0",
      },
      "status": "DONE",
      "travelTimePercentile": -1,
      "travelTimePercentiles": Array [
        50,
      ],
      "updatedAt": "2017-11-08T12:31:19.602Z",
      "updatedBy": "person@gmail.com",
      "variant": 0,
      "west": 20015,
      "width": 300,
      "workerVersion": "v3.2.0",
      "zoom": 9,
    },
    Object {
      "_id": "efgh",
      "accessGroup": "accessGroupName",
      "bounds": Object {},
      "bundleId": "1",
      "complete": true,
      "createdAt": "2017-11-08T12:30:50.133Z",
      "createdBy": "person@gmail.com",
      "cutoffMinutes": 60,
      "cutoffsMinutes": Array [
        60,
      ],
      "deleted": false,
      "destinationPointSetIds": Array [
        "Jobs_total",
      ],
      "grid": "Jobs_total",
      "height": 300,
      "name": "ANALYSIS 2",
      "nonce": "12345",
      "north": 20024,
      "projectId": "1",
      "request": Object {
        "accessModes": "WALK",
        "bikeSafe": 1,
        "bikeSlope": 1,
        "bikeSpeed": 4.166666666666667,
        "bikeTime": 1,
        "bounds": Object {
          "east": -76,
          "north": 39,
          "south": 38,
          "west": -77,
        },
        "bundleId": "1",
        "date": "2016-01-16",
        "defaultExpanded": true,
        "directModes": "WALK",
        "egressModes": "WALK",
        "fromLat": 12,
        "fromLon": 34,
        "fromTime": 25200,
        "maxBikeTime": 20,
        "maxRides": 4,
        "maxTripDurationMinutes": 60,
        "maxWalkTime": 20,
        "monteCarloDraws": 200,
        "name": "Profile Request Test",
        "projectId": "1",
        "reachabilityThreshold": 0,
        "regionId": "1",
        "toTime": 32400,
        "transitModes": "BUS,TRAM,RAIL,SUBWAY",
        "travelTimePercentile": 50,
        "variantIndex": 0,
        "walkSpeed": 1.3888888888888888,
        "workerVersion": "v4.0.0",
      },
      "status": "DONE",
      "travelTimePercentile": 50,
      "travelTimePercentiles": Array [
        50,
      ],
      "updatedAt": "2017-11-08T12:31:19.602Z",
      "updatedBy": "person@gmail.com",
      "variant": 1,
      "west": 20015,
      "width": 300,
      "workerVersion": "v3.2.0",
      "zoom": 9,
    },
  ],
  "type": "set regional analyses",
}
`;
